### PR TITLE
CSRE enhancements

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR adds some features on top of #295

### CODE REVIEW

**Comment by FabioLuporini** [comment]
Now rebased on master. 

Comments more than welcome

**Comment by FabioLuporini** [comment]
Some more fixes added

**Comment by FabioLuporini** [comment]
And now with stronger dle analysis and more tests

**Comment by FabioLuporini** [comment]
Excellent, Travis green.

Thoughts @mlange05 @navjotk  ?

File: tests/test_adjointA.py [hunk 1]
```diff
@@ -29,4 +29,4 @@ def test_acoustic(dimensions, time_order, space_order, fix_dim):
     term1 = np.dot(srca.data.reshape(-1), solver.source.data)
     term2 = linalg.norm(rec.data) ** 2
     print(term1, term2, ("%12.12f") % (term1 - term2), term1 / term2)
-    assert np.isclose(term1 / term2, 1.0, atol=0.001)
```
Comment (mlange05) [inline tests/test_adjointA.py]: What's happening here? Are we dropping accuracy somewhere?

File: examples/benchmark.py [hunk 2]
```diff
@@ -52,8 +52,8 @@
 
     devito = parser.add_argument_group("Devito")
     devito.add_argument("-dse", default="advanced", nargs="*",
-                        choices=["noop", "basic", "factorize", "approx-trigonometry",
-                                 "glicm", "advanced", "speculative"],
+                        choices=["noop", "basic", "advanced", "speculative",
+                                 "aggressive"],
```
Comment (mlange05) [inline examples/benchmark.py]: I realise these are in flux, but we will need some serious developer documentation for all these modes soon. Just a reminder though, not as part of this PR.

File: tests/test_adjointA.py (hunk 1 above)
Comment (FabioLuporini) [inline tests/test_adjointA.py]: It's because the receiver loops are now parallelised

File: examples/benchmark.py (hunk 2 above)
Comment (FabioLuporini) [inline examples/benchmark.py]: > Also, as mentioned, some developer-level docs are becoming more and more necessary to follow the flow of these optimisations. I know, low priority, but still important to keep in mind.

Yes, I'm planning to write it up while also writing the paper

File: tests/test_adjointA.py (hunk 1 above)
Comment (mlange05) [inline tests/test_adjointA.py]: That still seems like a lot. @mloubout can we live with this?

Comment (FabioLuporini) [inline tests/test_adjointA.py]: OK, so, I double checked and you're right, it turns out this change was actually unnecessary so I removed it.

The reason it was there was briefly discussed on Slack, when I was talking about parallelising the sources loop (not the receiver loops). 

So this is not a problem any more

Comment (mlange05) [inline tests/test_adjointA.py]: Excellent. Green light from me then. :smile:

**Comment by FabioLuporini** [comment]
@mlange05 OK, I've just rebased it on master, so I'll merge it as soon as Travis confirms his happiness

**Comment by FabioLuporini** [comment]
Merging...

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**